crates/constraint-vm/src/lib.rs

```diff
@@ -358,35 +350,39 @@ pub fn step_on_total_control_flow(
 }
 
 /// Step forward constraint checking by the given temporary operation.
-pub fn step_on_temporary(
-    op: asm::Temporary,
-    stack: &mut Stack,
-    memory: &mut Memory,
-) -> OpResult<()> {
+pub fn step_on_temporary(op: asm::Memory, stack: &mut Stack, memory: &mut Memory) -> OpResult<()> {
     match op {
-        asm::Temporary::Alloc => {
+        asm::Memory::Alloc => {
             let w = stack.pop()?;
             let len = memory.len()?;
             memory.alloc(w)?;
             Ok(stack.push(len)?)
         }
-        asm::Temporary::Store => {
+        asm::Memory::Store => {
             let [addr, w] = stack.pop2()?;
-            memory.store(addr, w)
+            memory.store(addr, w)?;
+            Ok(())
         }
-        asm::Temporary::Load => stack.pop1_push1(|addr| memory.load(addr)),
-        asm::Temporary::Free => {
+        asm::Memory::Load => stack.pop1_push1(|addr| {
+            let w = memory.load(addr)?;
+            Ok(w)
+        }),
+        asm::Memory::Free => {
             let addr = stack.pop()?;
-            memory.free(addr)
+            memory.free(addr)?;
+            Ok(())
         }
-        asm::Temporary::LoadRange => {
+        asm::Memory::LoadRange => {
             let [addr, size] = stack.pop2()?;
             let words = memory.load_range(addr, size)?;
             Ok(stack.extend(words)?)
         }
-        asm::Temporary::StoreRange => {
+        asm::Memory::StoreRange => {
             let addr = stack.pop()?;
-            stack.pop_len_words(|words| memory.store_range(addr, words))?;
+            stack.pop_len_words(|words| {
+                memory.store_range(addr, words)?;
+                Ok::<_, OpError>(())
+            })?;
```

Comment:
Looks like this is slightly buggy. We should get the data before the address. Something like:
```rust
        asm::Memory::StoreRange => {
            let words = stack.pop_len_words(|words| Ok::<_, OpSyncError>(words.to_vec()))?;
            let addr = stack.pop()?;
            memory.store_range(addr, &words)?;
            Ok(())
        }
```
Also, maybe we should add explicit tests that use `StoreRange` instead of just relying on testing `store_range`?

Reply:
I've addressed this in [this commit](https://github.com/essential-contributions/essential-base/pull/229/commits/e9491512fe578550470701e4f8e0aa6d5e539bd4) in https://github.com/essential-contributions/essential-base/pull/229, and the following commit adds some tests for the ops!